common/nexus/payload_serializer.go

```diff
@@ -0,0 +1,221 @@
+		err := payload.Unmarshal(content.Data)
+		if err != nil {
+			return err
+		}
+		return nil
+	case "application/json":
+		if params["format"] == "protobuf" {
+			payload.Metadata["encoding"] = []byte("json/protobuf")
+			messageType := params["message-type"]
+			if messageType != "" {
+				payload.Metadata["messageType"] = []byte(messageType)
+			}
+		} else {
+			payload.Metadata["encoding"] = []byte("json/plain")
+		}
+	case "application/x-protobuf":
+		payload.Metadata["encoding"] = []byte("binary/protobuf")
+		messageType := params["message-type"]
+		if messageType != "" {
+			payload.Metadata["messageType"] = []byte(messageType)
+		}
+	case "application/octet-stream":
+		payload.Metadata["encoding"] = []byte("binary/plain")
+	default:
+		// Should be unreachable.
+		return fmt.Errorf("%w: standard content detection failed for %q", errSerializer, mediaType)
+	}
+	return nil
+}
+
+func isStandardNexusContent(content *nexus.Content) bool {
+	h := content.Header
+	// We assume that encoding is handled by the transport layer and the content is decoded.
+	delete(h, "encoding")
+	// Length can safely be ignored.
+	delete(h, "length")
+
+	if len(h) > 1 {
+		return false
+	}
+	contentType := h.Get("type")
+	if contentType == "" {
+		return len(h) == 0 && len(content.Data) == 0
+	}
+	mediaType, params, err := mime.ParseMediaType(contentType)
+	if err != nil {
+		return false
+	}
+
+	switch mediaType {
+	case "application/octet-stream",
+		"application/x-temporal-payload":
+		return len(params) == 0
+	case "application/json":
+		if params["format"] == "protobuf" {
+			if params["message-type"] != "" {
+				return len(params) == 2
+			}
+			return len(params) == 1
+		}
+		return len(params) == 0
+	case "application/x-protobuf":
+		if params["message-type"] != "" {
+			return len(params) == 1
+		}
+		return len(params) == 0
+	}
+	return false
+}
+
+// Serialize implements nexus.Serializer.
+func (payloadSerializer) Serialize(v any) (*nexus.Content, error) {
+	payload, ok := v.(*commonpb.Payload)
+	if !ok {
+		return nil, fmt.Errorf("%w: cannot serialize %v", errSerializer, v)
+	}
+
+	if payload == nil {
+		return &nexus.Content{Header: nexus.Header{}}, nil
+	}
+
+	if !isStandardPayload(payload) {
+		data, err := payload.Marshal()
+		if err != nil {
+			return nil, fmt.Errorf("%w: payload marshal error: %w", errSerializer, err)
+		}
+		return &nexus.Content{
+			Header: nexus.Header{"type": "application/x-temporal-payload"},
+			Data:   data,
+		}, nil
+	}
+
+	content := nexus.Content{Header: nexus.Header{}, Data: payload.Data}
+	encoding := string(payload.Metadata["encoding"])
+	messageType := string(payload.Metadata["messageType"])
+
+	switch encoding {
+	case "unknown/nexus-content":
+		for k, v := range payload.Metadata {
+			if k != "encoding" {
+				content.Header[k] = string(v)
+			}
+		}
+	case "json/protobuf":
+		content.Header["type"] = fmt.Sprintf("application/json; format=protobuf")
+		if messageType != "" {
```

Comment:
I don't think any SDKs leave this off anymore and haven't for a while, I think we can require usage of a somewhat modern SDK version for use with Nexus protos

Reply:
I'm fine with that.